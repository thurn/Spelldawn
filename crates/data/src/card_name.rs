// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use strum_macros::Display;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display, Serialize, Deserialize)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}

impl PartialOrd<Self> for CardName {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for CardName {
    fn cmp(&self, other: &Self) -> Ordering {
        self.to_string().cmp(&other.to_string())
    }
}
