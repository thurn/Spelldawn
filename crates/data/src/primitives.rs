// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Fundamental types and data structures for Spelldawn

use std::fmt;

use serde::{Deserialize, Serialize};

pub type TurnNumber = u32;
pub type ActionCount = u32;
pub type ManaValue = u32;
pub type PointsValue = u32;
pub type HealthValue = u32;
pub type AttackValue = u32;
pub type ShieldValue = u32;
pub type BoostCount = u32;
pub type LevelValue = u32;

/// Identifies a player across different games
#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
pub struct PlayerId {
    pub value: u64,
}

impl PlayerId {
    pub fn new(value: u64) -> Self {
        Self { value }
    }

    /// Byte array representation of this ID
    pub fn key(&self) -> [u8; 8] {
        self.value.to_be_bytes()
    }
}

impl fmt::Debug for PlayerId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Identifies an ongoing game
#[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
pub struct GameId {
    pub value: u64,
}

impl GameId {
    pub fn new(value: u64) -> Self {
        Self { value }
    }

    /// Byte array representation of this ID
    pub fn key(&self) -> [u8; 8] {
        self.value.to_be_bytes()
    }
}

impl fmt::Debug for GameId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// The two players in a game: Overlord & Champion
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub enum Side {
    Overlord,
    Champion,
}

impl Side {
    /// Gets the opponent of the provided player
    pub fn opponent(&self) -> Self {
        match self {
            Side::Champion => Self::Overlord,
            Side::Overlord => Self::Champion,
        }
    }
}

impl fmt::Debug for Side {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Side::Overlord => "Overlord",
                Side::Champion => "Champion",
            }
        )
    }
}

/// Identifies a card in an ongoing game
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct CardId {
    pub side: Side,
    pub index: usize,
}

impl CardId {
    pub fn new(side: Side, index: usize) -> Self {
        Self { side, index }
    }
}

impl fmt::Debug for CardId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            match self.side {
                Side::Overlord => "O",
                Side::Champion => "C",
            },
            self.index
        )
    }
}

/// Identifies an ability position within a card's 'abilities' vector
#[derive(PartialEq, Eq, Hash, Copy, Clone, Ord, PartialOrd, Serialize, Deserialize)]
pub struct AbilityIndex(pub usize);

impl fmt::Debug for AbilityIndex {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// Identifies an ability within a card. Abilities are the only game entity
/// which may contain delegates..
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct AbilityId {
    pub card_id: CardId,
    pub index: AbilityIndex,
}

impl fmt::Debug for AbilityId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}[{:?}]", self.card_id, self.index)
    }
}

impl AbilityId {
    pub fn new(card_id: CardId, index: usize) -> Self {
        Self { card_id, index: AbilityIndex(index) }
    }
}

/// Uniquely identifies a raid within a given game
#[derive(PartialEq, Eq, Hash, Copy, Clone, Serialize, Deserialize)]
pub struct RaidId(pub u32);

impl fmt::Debug for RaidId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "%{}", self.0)
    }
}

/// Contains the URL of an image asset within a game
#[derive(PartialEq, Eq, Hash, Debug, Clone, Serialize, Deserialize)]
pub struct Sprite {
    pub address: String,
}

impl Sprite {
    pub fn new(address: impl Into<String>) -> Self {
        Self { address: address.into() }
    }
}

/// The schools of magic, which provide restrictions on players during
/// deckbuilding
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum School {
    Neutral,
    Shadow,
    Nature,
    Time,
}

/// The possible Rooms in which the Overlord player may play their cards.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum RoomId {
    /// The Overlord's deck
    Vault,
    /// The Overlord's hand
    Sanctum,
    /// The Overlord's discard pile
    Crypts,
    RoomA,
    RoomB,
    RoomC,
    RoomD,
    RoomE,
}

impl RoomId {
    /// An 'inner room' is one of the three predefined rooms for the Overlord
    /// player's deck, hand, and discard pile. Inner rooms cannot contain
    /// Schemes or Projects.
    pub fn is_inner_room(&self) -> bool {
        matches!(self, RoomId::Vault | RoomId::Sanctum | RoomId::Crypts)
    }
}

/// Used to control where a card is rendered within a room
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum RoomLocation {
    Defender,
    Occupant,
}

/// Used to control where an item is rendered within the Champion's item display
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum ItemLocation {
    Weapons,
    Artifacts,
}

/// The three possible factions of weapons and minions. Minions can only be
/// damaged by weapons from the same faction.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum Faction {
    Mortal,
    Abyssal,
    Infernal,
}

/// Rarity of a card, used to determine how likely it is to appear in randomized
/// rewards.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum Rarity {
    Common,
    Uncommon,
    Rare,
    Epic,

    /// Card cannot be obtained via random rewards
    None,
}

/// Possible types of cards
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum CardType {
    Spell,
    Weapon,
    Artifact,
    Minion,
    Project,
    Scheme,
    Upgrade,
    Identity,
}

/// Subtypes of cards
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Serialize, Deserialize)]
pub enum CardSubtype {
    Silvered,
}

/// Describes a boost ability activation
#[derive(PartialEq, Eq, Hash, Debug, Clone, Copy, Serialize, Deserialize)]
pub struct BoostData {
    /// Boosted card
    pub card_id: CardId,
    /// How many times was the boost applied?
    pub count: u32,
}
