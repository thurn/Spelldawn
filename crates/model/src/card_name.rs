// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use strum_macros::Display;

use convert_case::{Case, Casing};

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Display)]
pub enum CardName {
    /// Empty identity cards with no associated rules text, for use in unit tests
    TestChampionIdentity,
    TestOverlordIdentity,

    ArcaneRecovery,
    Greataxe,
    GoldMine,
    IceDragon,
    DungeonAnnex,
}

impl CardName {
    pub fn displayed_name(&self) -> String {
        format!("{}", self).from_case(Case::Pascal).to_case(Case::Title)
    }
}
