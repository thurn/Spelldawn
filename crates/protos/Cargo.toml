[package]
name = "protos"
version = "0.0.0"
description = "Generated protocol buffers code for the GRPC layer"
license = "Apache-2.0"
edition = "2021"

[lib]
doctest = false
test = false

[[bin]]
name = "gen_protos"
path = "src/gen_protos.rs"
test = false

[dependencies]
prost = "0.10.4"
tonic = { version = "0.7.2", features = ["compression"] }
tonic-build = { version = "0.7.2", features = ["compression"] }
