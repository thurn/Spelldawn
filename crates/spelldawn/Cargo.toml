[package]
name = "spelldawn"
version = "0.0.0"
authors = ["Spelldawn Team"]
homepage = "https://github.com/thurn/spelldawn"
description = "An asymmetric trading card game"
license = "Apache-2.0"
default-run = "spelldawn"
edition = "2021"
rust-version = "1.57"

[[bin]]
name = "spelldawn"
path = "src/main.rs"

[dependencies]
prost = "0.9.0"
prost-types = "0.9.0"
tokio = { version = "1.14.0", features = ["macros", "rt-multi-thread", "sync", "time"] }
tokio-stream = "0.1.8"
tonic = "0.6.1"
tonic-web = "0.2"

protos = { path = "../protos", version = "0.0.0" }