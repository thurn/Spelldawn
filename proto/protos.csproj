<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
    <Protobuf_NoWarnMissingExpected>true</Protobuf_NoWarnMissingExpected>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Grpc.Tools" Version="2.39.1">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

  <ItemGroup>
    <Protobuf Include="**/*.proto" OutputDir="%(RelativeDir)" CompileOutputs="false"  />
  </ItemGroup>

</Project>
