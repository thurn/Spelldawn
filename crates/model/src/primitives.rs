// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::card_name::CardName;

pub type TurnNumber = u32;
pub type ManaValue = u32;
pub type ActionCount = u32;
pub type Score = u32;
pub type HealthValue = u32;
pub type AttackValue = u32;
pub type ShieldValue = u32;
pub type BoostCount = u32;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub struct CardId {
    pub index: usize,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub struct EventId(pub u32);

/// Identifies an ability within a card. Abilities are the only game entity which may contain
/// delegates. Abilities are identified by their position within the card's 'abilities',
/// or 'activated_abilities' vector.
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone, Ord, PartialOrd)]
pub struct AbilityIndex(pub usize);

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub struct AbilityId {
    pub card_id: CardId,
    pub index: AbilityIndex,
}

impl AbilityId {
    pub fn new(card_id: CardId, index: usize) -> Self {
        Self { card_id, index: AbilityIndex(index) }
    }
}

/// Uniquely identifies a raid within a given game
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub struct RaidId(pub u32);

/// Identifies an encounter within a given raid
#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub struct EncounterId {
    pub raid_id: RaidId,
    pub step_id: u32,
}

#[derive(PartialEq, Eq, Hash, Debug, Clone)]
pub struct SpriteAddress(pub String);

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum Side {
    Champion,
    Overlord,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum School {
    Shadow,
    Nature,
    Time,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum RoomId {
    Treasury,
    Sanctum,
    Crypts,
    RoomA,
    RoomB,
    RoomC,
    RoomD,
    RoomE,
}

pub type DefenderIndex = u32;

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum RoomLocation {
    Defender(DefenderIndex),
    InRoom,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum ItemLocation {
    Weapons,
    Artifacts,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum Faction {
    Mortal,
    Abyssal,
    Infernal,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum Rarity {
    Common,
    Uncommon,
    Rare,
    Epic,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum CardType {
    Spell,
    Weapon,
    Minion,
    Project,
    Scheme,
    Identity,
    Token,
}

#[derive(PartialEq, Eq, Hash, Debug, Copy, Clone)]
pub enum CardSubtype {
    Silvered,
}
