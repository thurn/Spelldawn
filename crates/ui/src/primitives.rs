// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Contains primitive design elements such as colors & fonts. Intended to be
//! used via wildcard import in UI definition files.

use protos::spelldawn::{FlexColor, FontAddress};

pub const WHITE: Option<FlexColor> =
    Some(FlexColor { red: 1.0, green: 1.0, blue: 1.0, alpha: 1.0 });
pub const BLACK: Option<FlexColor> =
    Some(FlexColor { red: 0.0, green: 0.0, blue: 0.0, alpha: 1.0 });
pub const GRAY: Option<FlexColor> = Some(FlexColor { red: 0.5, green: 0.5, blue: 0.5, alpha: 1.0 });
pub const RED: Option<FlexColor> = Some(FlexColor { red: 1.0, green: 0.0, blue: 0.0, alpha: 1.0 });
pub const GREEN: Option<FlexColor> =
    Some(FlexColor { red: 0.0, green: 1.0, blue: 0.0, alpha: 1.0 });
pub const BLUE: Option<FlexColor> = Some(FlexColor { red: 0.0, green: 0.0, blue: 1.0, alpha: 1.0 });
pub const MAGENTA: Option<FlexColor> =
    Some(FlexColor { red: 1.0, green: 0.0, blue: 1.0, alpha: 1.0 });
pub const CYAN: Option<FlexColor> = Some(FlexColor { red: 0.0, green: 1.0, blue: 1.0, alpha: 1.0 });
pub const YELLOW: Option<FlexColor> =
    Some(FlexColor { red: 1.0, green: 1.0, blue: 0.0, alpha: 1.0 });

/// Possible interface colors
#[derive(Debug, Clone, Copy)]
pub enum Color {
    TitleText,
    ButtonLabel,
}

/// Returns the [FlexColor] to use for a given interface color. Prefer using
/// this function to directly accessing the color constants in this module.
pub fn color(color: Color) -> Option<FlexColor> {
    match color {
        Color::TitleText => WHITE,
        Color::ButtonLabel => WHITE,
    }
}

/// Possible interface fonts
#[derive(Debug, Clone, Copy)]
pub enum Font {
    Default,
}

pub fn font(font: Font) -> Option<FontAddress> {
    Some(FontAddress {
        address: match font {
            Font::Default => "Fonts/Roboto",
        }
        .to_string(),
    })
}
