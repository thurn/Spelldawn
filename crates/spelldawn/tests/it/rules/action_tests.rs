// Copyright Â© Spelldawn 2021-present

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

//    https://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use data::card_name::CardName;
use data::primitives::Side;
use insta::assert_debug_snapshot;
use protos::spelldawn::game_action::Action;
use protos::spelldawn::{
    card_target, CardTarget, ClientRoomLocation, DrawCardAction, GainManaAction, PlayCardAction,
    PlayerName,
};
use test_utils::*;

#[test]
fn connect() {
    let mut g =
        new_game(Side::Overlord, Args { id_basis: Some(2), connect: false, ..Args::default() });
    let response = g.connect(g.user_id(), Some(g.game_id()));
    assert_ok(&response);
    assert_debug_snapshot!(response);
}

#[test]
fn connect_to_ongoing() {
    let mut g = new_game(
        Side::Overlord,
        Args {
            id_basis: Some(4),
            actions: 3,
            next_draw: Some(CardName::IceDragon),
            ..Args::default()
        },
    );
    let r1 = g.connect(g.user_id(), Some(g.game_id()));
    assert_ok(&r1);
    let r2 = g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id());
    assert_identical(vec![CardName::IceDragon], g.user.cards.hand(PlayerName::User));
    assert_ok(&r2);
    let r3 = g.connect(g.opponent_id(), Some(g.game_id()));
    assert_ok(&r3);
    assert_debug_snapshot!(r3);
}

#[test]
fn draw_card() {
    let mut g = new_game(
        Side::Overlord,
        Args {
            id_basis: Some(6),
            actions: 3,
            next_draw: Some(CardName::IceDragon),
            ..Args::default()
        },
    );
    let response = g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id());
    assert_identical(vec![CardName::IceDragon], g.user.cards.hand(PlayerName::User));
    assert_eq!(vec![HIDDEN_CARD], g.opponent.cards.hand(PlayerName::Opponent));
    assert_eq!(2, g.user().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_commands_match(
        &response,
        vec![
            "UpdateGameView",     // Spend mana & actions
            "CreateOrUpdateCard", // Create card on top of deck
            "MoveGameObjects",    // Move card to hand
        ],
    );
    assert_debug_snapshot!(response);
}

#[test]
fn cannot_draw_card_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.opponent_id()));
}

#[test]
fn cannot_draw_when_out_of_action_points() {
    let mut g = new_game(Side::Overlord, Args { actions: 0, ..Args::default() });
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id()));
}

#[test]
fn cannot_draw_during_raid() {
    let mut g = new_game(
        Side::Overlord,
        Args { raid: Some(TestRaid { priority: Side::Overlord }), ..Args::default() },
    );
    assert_error(g.perform_action(Action::DrawCard(DrawCardAction {}), g.user_id()));
}

#[test]
fn play_card() {
    let mut g = new_game(
        Side::Champion,
        Args { id_basis: Some(8), actions: 3, mana: 5, ..Args::default() },
    );
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    let response = g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    );
    assert_eq!(2, g.user().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(9, g.user().mana());
    assert_eq!(9, g.opponent.other_player.mana());
    assert_identical(vec![CardName::ArcaneRecovery], g.user.cards.discard_pile(PlayerName::User));
    assert_identical(
        vec![CardName::ArcaneRecovery],
        g.opponent.cards.discard_pile(PlayerName::Opponent),
    );
    assert_commands_match_lists(
        &response,
        vec![
            "UpdateGameView",     // Spend mana & actions
            "CreateOrUpdateCard", // Update canPlay
            "MoveGameObjects",    // Move to discard
        ],
        // Update card and move it to staging for a short delay, then move it to discard
        vec!["UpdateGameView", "CreateOrUpdateCard", "MoveGameObjects", "Delay", "MoveGameObjects"],
    );
    assert_debug_snapshot!(response);
}

#[test]
fn play_hidden_card() {
    let mut g = new_game(
        Side::Overlord,
        Args { id_basis: Some(10), actions: 3, mana: 0, ..Args::default() },
    );
    let card_id = g.add_to_hand(CardName::DungeonAnnex);
    let response = g.perform_action(
        Action::PlayCard(PlayCardAction {
            card_id: Some(card_id),
            target: Some(CardTarget {
                card_target: Some(card_target::CardTarget::RoomId(CLIENT_ROOM_ID.into())),
            }),
        }),
        g.user_id(),
    );
    assert_eq!(2, g.user().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(0, g.user().mana());
    assert_eq!(0, g.opponent.other_player.mana());
    assert_identical(
        vec![CardName::DungeonAnnex],
        g.user.cards.room_cards(ROOM_ID, ClientRoomLocation::Back),
    );
    assert_eq!(vec![HIDDEN_CARD], g.opponent.cards.room_cards(ROOM_ID, ClientRoomLocation::Back));
    assert_commands_match_lists(
        &response,
        // Update state and move to room
        vec!["UpdateGameView", "CreateOrUpdateCard", "MoveGameObjects"],
        // No card update required, state does not change
        vec!["UpdateGameView", "MoveGameObjects"],
    );
    assert_debug_snapshot!(response);
}

#[test]
fn cannot_play_card_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn cannot_play_card_when_out_of_action_points() {
    let mut g = new_game(Side::Champion, Args { actions: 0, ..Args::default() });
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn cannot_play_card_during_raid() {
    let mut g = new_game(
        Side::Champion,
        Args { raid: Some(TestRaid { priority: Side::Overlord }), ..Args::default() },
    );
    let card_id = g.add_to_hand(CardName::ArcaneRecovery);
    assert_error(g.perform_action(
        Action::PlayCard(PlayCardAction { card_id: Some(card_id), target: None }),
        g.user_id(),
    ));
}

#[test]
fn gain_mana() {
    let mut g = new_game(
        Side::Overlord,
        Args { id_basis: Some(12), actions: 3, mana: 5, ..Args::default() },
    );
    let response = g.perform_action(Action::GainMana(GainManaAction {}), g.user_id());
    assert_eq!(2, g.user().actions());
    assert_eq!(2, g.opponent.other_player.actions());
    assert_eq!(6, g.user().mana());
    assert_eq!(6, g.opponent.other_player.mana());
    assert_commands_match(&response, vec!["UpdateGameView"]);
    assert_debug_snapshot!(response);
}

#[test]
fn cannot_gain_mana_on_opponent_turn() {
    let mut g = new_game(Side::Overlord, Args::default());
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.opponent_id()));
}

#[test]
fn cannot_gain_mana_when_out_of_action_points() {
    let mut g = new_game(Side::Overlord, Args { actions: 0, ..Args::default() });
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()));
}

#[test]
fn cannot_gain_mana_during_raid() {
    let mut g = new_game(
        Side::Overlord,
        Args { raid: Some(TestRaid { priority: Side::Overlord }), ..Args::default() },
    );
    assert_error(g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()));
}

#[test]
fn switch_turn() {
    let mut g = new_game(Side::Overlord, Args { actions: 3, mana: 5, ..Args::default() });
    g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()).unwrap();
    g.perform_action(Action::GainMana(GainManaAction {}), g.user_id()).unwrap();
    let response = g.perform_action(Action::GainMana(GainManaAction {}), g.user_id());
    assert_eq!(8, g.user().mana());
    assert_eq!(8, g.opponent.other_player.mana());
    assert_eq!(0, g.user().actions());
    assert_eq!(0, g.opponent.other_player.actions());
    assert_eq!(3, g.user.other_player.actions());
    assert_eq!(3, g.opponent.this_player.actions());
    assert_eq!(g.user.data.priority(), PlayerName::Opponent);
    assert_eq!(g.opponent.data.priority(), PlayerName::User);
    assert_commands_match(&response, vec!["UpdateGameView", "DisplayGameMessage"]);
}
