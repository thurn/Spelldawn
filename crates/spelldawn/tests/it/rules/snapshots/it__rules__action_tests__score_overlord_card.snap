---
source: crates/spelldawn/tests/it/rules/action_tests.rs
assertion_line: 233
expression: "Summary::run(&response)"

---

command_list: 
    SetMusic: 
        music_state: Silent
    PlaySound: 
        sound: "<AudioClipAddress>"
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 3
            position: ObjectPositionScoreAnimation
    PlayEffect: 
        effect: "<EffectAddress>"
        position: O45
        duration: 700
        sound: "<AudioClipAddress>"
    PlayEffect: 
        effect: "<EffectAddress>"
        position: O45
        duration: 300
        sound: "<AudioClipAddress>"
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 3
            position: ObjectPositionIdentity { owner: User }
    MoveGameObjects: 
        id: Identity(User)
        position: 
            sorting_key: 0
            position: ObjectPositionIdentityContainer { owner: User }
    UpdateGameView: 
        user: 
            side: Overlord
            mana: 7
            action_tracker: 1
            score: 1
        current_priority: User
        raid_active: false
channel_response: 
    VisitRoom: 
        initiator: Opponent
        room_id: RoomA
        visit_type: LevelUpRoom
    Delay: 500
    CreateOrUpdateCard: 
        card: 
            card_id: O45
            revealed_to_viewer: true
            revealed_to_opponent: true
            card_icons: 
                arena_icon: "3"
            arena_frame: "<SpriteAddress>"
            owning_player: Opponent
            revealed_card: 
                card_frame: "<SpriteAddress>"
                title_background: "<SpriteAddress>"
                jewel: "<SpriteAddress>"
                image: "<SpriteAddress>"
                title: "Test Scheme 31"
                rules_text: "<RulesText>"
                targeting: "PickRoom"
                on_release_position: 
                    sorting_key: 4294967295
                    position: ObjectPositionRoom { room_id: Unspecified, room_location: Back }
                can_play: false
        create_position: 
            sorting_key: 3
            position: ObjectPositionIdentity { owner: Opponent }
        create_animation: Unspecified
    SetMusic: 
        music_state: Silent
    PlaySound: 
        sound: "<AudioClipAddress>"
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 3
            position: ObjectPositionScoreAnimation
    PlayEffect: 
        effect: "<EffectAddress>"
        position: O45
        duration: 700
        sound: "<AudioClipAddress>"
    PlayEffect: 
        effect: "<EffectAddress>"
        position: O45
        duration: 300
        sound: "<AudioClipAddress>"
    MoveGameObjects: 
        id: O45
        position: 
            sorting_key: 3
            position: ObjectPositionIdentity { owner: Opponent }
    MoveGameObjects: 
        id: Identity(Opponent)
        position: 
            sorting_key: 0
            position: ObjectPositionIdentityContainer { owner: Opponent }
    UpdateGameView: 
        opponent: 
            side: Overlord
            mana: 7
            action_tracker: 1
            score: 1
        current_priority: Opponent
        raid_active: false
    Delay: 1500
